pandas/core/reshape/merge.py

```diff
@@ -1661,6 +1663,13 @@ def _get_merge_keys(self):
                     )
                 raise MergeError(msg)
 
+            if i >= len_by_cols:
```

Comment:
no, we don't want to do this.

put your changes in the super call to _get_merge_keys. just do it upfront (that is what was not clear before).

Reply:
I think you mean _maybe_coerce_merge_keys then? The dtype validation for the regular merge is done there. 

The _get_merge_keys function I changed is already from _AsOfMerge, so I could not overwrite this again. This function overwrites the original _get_merge_keys function and calls the super() function before checking if both join cols have the same dtype